### PR TITLE
[ORCA] Alternative approach for comparison expression reordering

### PR DESCRIPTION
**Pull request by dgkimura**
An effect of the preprocessor step added in commit https://github.com/greenplum-db/gpdb/commit/063b0abc3cccb488f23060954f959551e4c2f544, is that
compare expressions can be added that are of the form (CONST OP IDENT).
Issue with that is that when the range of a constraint interval is
constructed it assumes a specific order of the operands such that the
IDENT is on the left side.

There are 2 different approaches to solve that issue:

  1) Update constraint constructor to use the commutative operator
  2) Fix up the tree to reorder the scalar compare children

This commit goes with the second approach by leveraging preprocessor
step PexprReorderScalarCmpChildren(). It includes a partial revert of
commmit https://github.com/greenplum-db/gpdb/commit/25fc6e003d9be02c03ec6e2333dcbb8a73e7b828, but keeps the added testcases. It also removes the
code added for first approach that was introduced in commit https://github.com/greenplum-db/gpdb/commit/063b0abc3cccb488f23060954f959551e4c2f544.

### CODE REVIEW

**Review by gpopt — COMMENTED** [review]
Is the step 26 removable? I am not sure, since we already reordered the scalar cmp operator here in step 7.b.

**Comment by gpopt** [comment]
This looks better than previous change.

**Comment by dgkimura** [comment]
On Wednesday, March 22, 2023 1:19 PM gpopt wrote:
> Is the step 26 removable? I am not sure, since we already reordered the scalar cmp operator here in step 7.b.

I think you're right that it seems removable.  But it's a little hard to tell because there are so many usages of scalar cmp with different flavors.  I skimmed through the reordered steps and I didn't find anything that I would expect to create a (const op ident) scalar cmp.

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 1]
```diff
@@ -3314,19 +3314,25 @@ CExpressionPreprocessor::PexprPreprocess(
...
-	// (7) substitute constant predicates
+	// (7.a) substitute constant predicates
 	ExprToConstantMap *phmExprToConst = GPOS_NEW(mp) ExprToConstantMap(mp);
 	CExpression *pexprPredWithConstReplaced =
 		PexprReplaceColWithConst(mp, pexprTrimmed2, phmExprToConst, true);
 	GPOS_CHECK_ABORT;
 	phmExprToConst->Release();
 	pexprTrimmed2->Release();
 
+	// (7.b) reorder the children of scalar cmp operator to ensure that left child is scalar ident and right child is scalar const
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3325]: limit the column width

**Comment by gpopt** [comment]
Given the following query:
```
SELECT * FROM qp_select WHERE 20 > a AND a >= 0;
```

If there is no `<` operator (by removing its metadata from DXL file, or suppose it is a user defined operator), will Orca generate correct plan?

**Comment by dgkimura** [comment]
On Thursday, March 23, 2023 1:52 PM gpopt wrote:
> Given the following query:
> 
> ```
> SELECT * FROM qp_select WHERE 20 > a AND a >= 0;
> ```
> 
> If there is no `<` operator (by removing its metadata from DXL file, or suppose it is a user defined operator), in other words, we can't reorder the expression, will Orca generate correct plan?

Yeah.  If there is no commutator operator then the reorder step is a no-op.

```Algebrized query:                                              
+--CLogicalSelect                                              
   |--CLogicalGet "qp_select" ("qp_select"), Columns: ["a" (0), ...
   +--CScalarBoolOp (EboolopAnd)
      |--CScalarCmp (>)                                        
      |  |--CScalarConst (20)                                  
      |  +--CScalarIdent "a" (0)
      +--CScalarCmp (>=)                                       
         |--CScalarIdent "a" (0)
         +--CScalarConst (0)                                   

Algebrized preprocessed query: 
+--CLogicalSelect                                              
   |--CLogicalGet "qp_select" ("qp_select"), Columns: ["a" (0), ...
   +--CScalarBoolOp (EboolopAnd)
      |--CScalarCmp (>)                                        
      |  |--CScalarConst (20)                                  
      |  +--CScalarIdent "a" (0)
      +--CScalarCmp (>=)                                       
         |--CScalarIdent "a" (0)
         +--CScalarConst (0)
         ```
         
And I think the plan is as you'd expect:
```
Physical plan: 
+--CPhysicalMotionGather(master)
   +--CPhysicalFilter
      |--CPhysicalTableScan "qp_select" ("qp_select")
      +--CScalarBoolOp (EboolopAnd)
         |--CScalarCmp (>) 
         |  |--CScalarConst (20)
         |  +--CScalarIdent "a" (0)
         +--CScalarCmp (>=)
            |--CScalarIdent "a" (0)
            +--CScalarConst (0)
```

**Comment by gpopt** [comment]
Can we add this as a minidump test case?

**Review by gpopt — APPROVED** [review]
Nice work!

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 2]
```diff
@@ -3314,19 +3314,26 @@ CExpressionPreprocessor::PexprPreprocess(
...
-	// (7) substitute constant predicates
+	// (7.a) substitute constant predicates
 	ExprToConstantMap *phmExprToConst = GPOS_NEW(mp) ExprToConstantMap(mp);
 	CExpression *pexprPredWithConstReplaced =
 		PexprReplaceColWithConst(mp, pexprTrimmed2, phmExprToConst, true);
 	GPOS_CHECK_ABORT;
 	phmExprToConst->Release();
 	pexprTrimmed2->Release();
 
+	// (7.b) reorder the children of scalar cmp operator to ensure that left
```
Comment (THANATOSLAVA) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3325]: Hmmm step 26 bubbled up to step 8. I wonder if any expression generated by old steps 8 through 25 could use the reordering, that necessitated the step to happen no earlier than 26? I imagine some steps are completely orthogonal, whereas some have to come in a certain sequence?

Comment (THANATOSLAVA) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3325]: Oh I missed that comment. Thanks! Can we add this to the comment of 7.b?

>(7.a) substitute constant predicates" happen before any normalization steps (e.g. "(11) infer predicates from constraints") and "(7.b) reorder the children of scalar cmp operator" must happen after 7.a.

**Review by THANATOSLAVA — APPROVED** [review]

**Comment by dgkimura** [comment]
On Thursday, March 23, 2023 5:17 PM gpopt wrote:
> Can we add this as a minidump test case?

Yes, `PushConstantSelectPredicateThruJoin-17.mdp` should handle that scenario:.

```sql
    -- Text starts_with operator ^@ doesnt' have commutator operator.
    CREATE TABLE t(a text);
    INSERT INTO t VALUES ('a'), ('b'), ('c');
    EXPLAIN (COSTS OFF) SELECT * FROM t WHERE 'a' ^@ a;
                    QUERY PLAN
    ------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)
       ->  Seq Scan on t
             Filter: ('a'::text ^@ a)
     Optimizer: Pivotal Optimizer (GPORCA)
```

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 2 above)
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3325]: I'll reword the comment.  But I'd rather exclude hardcoded numbers (e.g. 11) as I think it's not uncommon for us to reorder or add steps.  In which case keeping everywhere updated would be a headache.

**Comment by dgkimura** [comment]
Merged. Thanks for reviews!

**Pull request merged**